In [48]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [7]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv('../Credit_Risk/Resources/lending_data.csv')

# Review the DataFrame
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [53]:
# Separate the data into labels and features
X = df.copy()
X.drop("loan_status", axis=1, inplace= True)

# Separate the y variable, the labels
y = df['loan_status'].ravel()

# Separate the X variable, the features
# YOUR CODE HERE!

In [54]:
# Review the y variable Series
y[:5]

array([0, 0, 0, 0, 0], dtype=int64)

In [10]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [59]:
# Check the balance of our target values
pd.value_counts(y)

0    75036
1     2500
dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [13]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)



---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [92]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [64]:
# Make a prediction using the testing data
testing_predictions = logistic_regression_model.predict(X_test)
# dataframe with the actuals vs the predicted values
pd.DataFrame(
    {
        "Actuals": y_test,
        "Predictions": testing_predictions,
    }
).head(12)

Actuals  Predictions
0         0            0
1         0            0
2         0            0
3         0            0
4         0            0
5         0            0
6         0            0
7         0            0
8         0            0
9         0            0
10        1            1
11        0            0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [89]:
# Print the balanced_accuracy score of the model

balanced_accuracy_score(y_test, testing_predictions)


0.9520479254722232

In [97]:
print(logistic_regression_model.score(X_train, y_train))
print(logistic_regression_model.score(X_test, y_test))

0.9947308560359688
0.9938093272802311


In [108]:
# Import the model for sklearn's confusion matrix
from sklearn.metrics import confusion_matrix

# Print the confusion matrix for the training data
# print(training_matrix)
confusion_matrix(y_test, testing_predictions)

array([[18663,   102],
       [   56,   563]], dtype=int64)

In [112]:

# Create and save the testing classification report
testing_report = classification_report(y_test, testing_predictions)

# Print the testing classification report
print(testing_report)


              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The precision for predicting healthy loans is at 100%, and 85% for risky loans. The model is better at predicting healthy loans than risky ones.
Recall is at 99% for healthy loans and 91% for risky ones. Again the model shows performing best to predict false positives than false negatives. For F1 scores, healthy loans have an ideal score of 100%, followed by high risk loans at a score of 88%.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [78]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_ros, y_ros = ros.fit_resample(X_train,y_train)

len(y_ros)

112542

In [79]:
import array
# Count the distinct values of the resampled labels data
#np.unique(y_train, return_counts= True)
pd.value_counts(y_ros)

0    56271
1    56271
dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [80]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
ros = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
ros.fit(X_ros, y_ros)

# Make a prediction using the testing data
ros_predict = ros.predict(X_test)
ros_predict[:5]

array([0, 0, 0, 0, 0], dtype=int64)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [84]:
# Print the balanced_accuracy score of the model 
print(ros.score(X_train, y_train))
print(ros.score(X_test, y_test))

0.9947308560359688
0.9938093272802311


In [85]:
# Generate a confusion matrix for the model
confusion_matrix(y_test,ros_predict)

array([[18649,   116],
       [    4,   615]], dtype=int64)

In [86]:
# Print the classification report for the model
print(classification_report(y_test, ros_predict))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model, fit with oversampled data retained the level of precision, recall and f1-score of the original model for the healthy loans, 100%, 99% and 100% respectively.
We are noticing overall improvement on the results for the risky loans. However, the precision level lost 1 point at 84%. The recall is at 99% and f1-score at 91% which increased from previous fit.